File: go-rewrite/internal/api/weebcentral.go
// Weeb Central (weebcentral.com) source — an HTMX site (no JSON API), so we
// scrape its HTML fragments. Large, current scanlation library.

package api

import (
	"fmt"
	"html"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/devcheckra1n/manga-cli/go-rewrite/internal/util"
)

const (
	wcBase = "https://weebcentral.com"
	wcTTL  = time.Hour
)

var wcHeaders = map[string]string{"Referer": wcBase + "/"}

var (
	wcCoverRe  = regexp.MustCompile(`src="(https://[^"]*/cover/[^"]*/([A-Z0-9]{10,})\.[a-z]+)"`)
	wcSeriesRe = regexp.MustCompile(`href="https://weebcentral\.com/series/([A-Z0-9]{10,})/[^"]*"[^>]*>([^<]{1,160})<`)
	wcTitleRe  = regexp.MustCompile(`(?i)<title>(.+?)(?:\s*\|\s*Weeb Central)?</title>`)
	wcStripRe  = regexp.MustCompile(`(?i)\b(Manhwa|Manhua|Webtoon|Long Strip)\b`)
	wcSvgRe    = regexp.MustCompile(`(?s)<svg.*?</svg>`)
	// Labels look like "Chapter 385", "Episode 5", or "# 122".
	wcChapterRe = regexp.MustCompile(`(?s)href="https://weebcentral\.com/chapters/([A-Z0-9]{10,})".{0,600}?>\s*(?:Chapter\s+|Episode\s+|#\s*)([0-9]+(?:\.[0-9]+)?)\s*<`)
	wcImageRe   = regexp.MustCompile(`(?i)src="(https://[^"]+\.(?:webp|jpg|jpeg|png|avif))"`)
)

type weebcentralSource struct{}

func (weebcentralSource) ID() SourceID    { return SrcWeebCentral }
func (weebcentralSource) Label() string   { return "Weeb Central (weebcentral.com)" }
func (weebcentralSource) Available() bool { return true }

func wcFragment(path string) (string, error) {
	return httpText(wcBase+path, wcHeaders)
}

// wcParseSeries parses weebcentral series cards (search + discovery fragments).
func wcParseSeries(doc string) []SearchResult {
	coverByID := map[string]string{}
	for _, m := range wcCoverRe.FindAllStringSubmatch(doc, -1) {
		coverByID[m[2]] = m[1]
	}
	var out []SearchResult
	seen := map[string]bool{}
	for _, m := range wcSeriesRe.FindAllStringSubmatch(doc, -1) {
		id := m[1]
		title := strings.TrimSpace(html.UnescapeString(m[2]))
		if title == "" || seen[id] {
			continue
		}
		seen[id] = true
		poster := coverByID[id]
		if poster == "" {
			poster = "https://temp.compsci88.com/cover/normal/" + id + ".webp"
		}
		out = append(out, SearchResult{
			ID: id, Title: title, Poster: poster, Type: "Manga", Source: SrcWeebCentral,
		})
	}
	return out
}

func (weebcentralSource) Search(query string, _ SearchOpts) ([]SearchResult, error) {
	key := "wc:search:" + query
	var cached []SearchResult
	if util.ReadCache(util.SearchCache, key, 10*time.Minute, &cached) {
		return cached, nil
	}
	params := url.Values{}
	params.Set("text", query)
	params.Set("sort", "Best Match")
	params.Set("order", "Descending")
	params.Set("official", "Any")
	params.Set("display_mode", "Full Display")
	doc, err := wcFragment("/search/data?" + params.Encode())
	if err != nil {
		return nil, err
	}
	results := wcParseSeries(doc)
	if len(results) > 24 {
		results = results[:24]
	}
	util.WriteCache(util.SearchCache, key, results)
	return results, nil
}

func (weebcentralSource) Discovery(kind DiscoveryKind, page int, _ bool) ([]DiscoveryItem, error) {
	path := "/hot-series"
	switch kind {
	case RecentlyAdded:
		path = fmt.Sprintf("/recently-added/%d", page+1)
	case RecentlyUpdated:
		path = fmt.Sprintf("/latest-updates/%d", page+1)
	}
	doc, err := wcFragment(path)
	if err != nil {
		return nil, err
	}
	results := wcParseSeries(doc)
	if len(results) > 32 {
		results = results[:32]
	}
	out := make([]DiscoveryItem, len(results))
	for i, r := range results {
		out[i] = DiscoveryItem{ID: r.ID, Title: r.Title, Poster: r.Poster, Type: r.Type, Source: SrcWeebCentral}
	}
	return out, nil
}

// No public genre listing; let another source provide genres.
func (weebcentralSource) Filters() (Filters, error) {
	return Filters{Genres: []Genre{}, Tags: []Genre{}, Types: []Genre{}, Statuses: []Genre{}}, nil
}
func (weebcentralSource) Browse(BrowseFilter) ([]SearchResult, error) {
	return []SearchResult{}, nil
}

func (weebcentralSource) Info(seriesID string) (*MangaInfo, error) {
	key := "wc:info:" + seriesID
	var cached MangaInfo
	if util.ReadCache(util.MangaCache, key, wcTTL, &cached) {
		return &cached, nil
	}

	page, err := wcFragment("/series/" + seriesID)
	if err != nil {
		return nil, err
	}
	title := "Untitled"
	if m := wcTitleRe.FindStringSubmatch(page); m != nil {
		title = strings.TrimSpace(html.UnescapeString(m[1]))
	}
	longStrip := wcStripRe.MatchString(page)

	// Strip the inline SVG icons first — they're huge and contain `#color`
	// codes that would otherwise be mistaken for "# 122"-style chapter labels.
	listDoc, err := wcFragment("/series/" + seriesID + "/full-chapter-list")
	if err != nil {
		return nil, err
	}
	listDoc = wcSvgRe.ReplaceAllString(listDoc, "")

	type rawCh struct {
		id  string
		num float64
	}
	var raw []rawCh
	seen := map[string]bool{}
	for _, m := range wcChapterRe.FindAllStringSubmatch(listDoc, -1) {
		if seen[m[1]] {
			continue
		}
		seen[m[1]] = true
		n, _ := strconv.ParseFloat(m[2], 64)
		raw = append(raw, rawCh{m[1], n})
	}
	sort.SliceStable(raw, func(a, b int) bool { return raw[a].num < raw[b].num })

	typ := "Manga"
	if longStrip {
		typ = "Manhwa"
	}
	info := MangaInfo{ID: seriesID, Title: title, Type: typ, ForceStrip: longStrip}
	for i, ch := range raw {
		info.Chapters = append(info.Chapters, Chapter{
			ID: ch.id, Title: "Chapter " + trimFloat(ch.num), Number: ch.num, Index: i,
		})
	}
	util.WriteCache(util.MangaCache, key, info)
	return &info, nil
}

func (weebcentralSource) Pages(_ string, chapterID string) (*ReadChapter, error) {
	key := "wc:pages:" + chapterID
	var cached ReadChapter
	if util.ReadCache(util.ChaptersCache, key, wcTTL, &cached) {
		return &cached, nil
	}
	doc, err := wcFragment("/chapters/" + chapterID + "/images?is_prev=False&current_page=1&reading_style=long_strip")
	if err != nil {
		return nil, err
	}
	out := ReadChapter{ID: chapterID}
	for i, m := range wcImageRe.FindAllStringSubmatch(doc, -1) {
		out.Pages = append(out.Pages, Page{
			ID: fmt.Sprintf("%s-%d", chapterID, i), URL: m[1], Number: i, AspectRatio: 1,
		})
	}
	util.WriteCache(util.ChaptersCache, key, out)
	return &out, nil
}

func (weebcentralSource) Related(string, int, bool) ([]DiscoveryItem, error) {
	return []DiscoveryItem{}, nil
}

File: src/api/sources/weebcentral.ts
// Weeb Central (weebcentral.com) source — an HTMX site (no JSON API), so we scrape
// its HTML fragments. It has a large, current scanlation library (e.g. Berserk up
// to the latest chapter), which makes it a strong backup for titles MangaDex has
// delicensed.

import { httpText } from "../client.ts";
import { readCache, writeCache } from "../../utils/cache.ts";
import { MANGA_CACHE, SEARCH_CACHE, CHAPTERS_CACHE } from "../../utils/paths.ts";
import type { DiscoveryItem, Filters, MangaInfo, ReadChapter, SearchResult, Source } from "../types.ts";

const BASE = "https://weebcentral.com";
const HEADERS = { Referer: BASE + "/" };
const TTL = 60 * 60 * 1000;

function decodeEntities(s: string): string {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&quot;/g, '"')
    .replace(/&#0?39;/g, "'")
    .replace(/&#0?34;/g, '"')
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .trim();
}

/** Parse weebcentral series cards (used by search + discovery fragments). */
function parseSeries(html: string): SearchResult[] {
  const coverById = new Map<string, string>();
  for (const m of html.matchAll(/src="(https:\/\/[^"]*\/cover\/[^"]*\/([A-Z0-9]{10,})\.[a-z]+)"/g)) {
    coverById.set(m[2], m[1]);
  }
  const out: SearchResult[] = [];
  const seen = new Set<string>();
  for (const m of html.matchAll(
    /href="https:\/\/weebcentral\.com\/series\/([A-Z0-9]{10,})\/[^"]*"[^>]*>([^<]{1,160})</g,
  )) {
    const id = m[1];
    const title = decodeEntities(m[2]);
    if (!title || seen.has(id)) continue;
    seen.add(id);
    out.push({
      id,
      title,
      poster: coverById.get(id) ?? `https://temp.compsci88.com/cover/normal/${id}.webp`,
      type: "Manga",
      isAdult: false,
      source: "weebcentral",
    });
  }
  return out;
}

async function fragment(path: string): Promise<string> {
  return httpText(BASE + path, { headers: HEADERS });
}

export const weebcentral: Source = {
  id: "weebcentral",
  label: "Weeb Central (weebcentral.com)",
  available: true,

  async search(query): Promise<SearchResult[]> {
    const key = `wc:search:${query}`;
    const hit = await readCache<SearchResult[]>(SEARCH_CACHE, key, 10 * 60 * 1000);
    if (hit) return hit;
    const params = new URLSearchParams({
      text: query,
      sort: "Best Match",
      order: "Descending",
      official: "Any",
      display_mode: "Full Display",
    });
    const html = await fragment(`/search/data?${params.toString()}`);
    const results = parseSeries(html).slice(0, 24);
    await writeCache(SEARCH_CACHE, key, results);
    return results;
  },

  async discovery(kind, page): Promise<DiscoveryItem[]> {
    const path =
      kind === "recentlyAdded"
        ? `/recently-added/${page + 1}`
        : kind === "recentlyUpdated"
          ? `/latest-updates/${page + 1}`
          : `/hot-series`;
    return parseSeries(await fragment(path)).slice(0, 32);
  },

  // No public genre listing; let another source provide genres.
  async filters(): Promise<Filters> {
    return { genres: [], tags: [], types: [], statuses: [] };
  },
  async browse(): Promise<SearchResult[]> {
    return [];
  },

  async info(seriesId): Promise<MangaInfo> {
    const key = `wc:info:${seriesId}`;
    const hit = await readCache<MangaInfo>(MANGA_CACHE, key, TTL);
    if (hit) return hit;

    const page = await fragment(`/series/${seriesId}`);
    const titleMatch = page.match(/<title>([^<]+?)(?:\s*\|\s*Weeb Central)?<\/title>/i);
    const title = titleMatch ? decodeEntities(titleMatch[1]) : "Untitled";
    const longStrip = /\b(Manhwa|Manhua|Webtoon|Long Strip)\b/i.test(page);

    // Strip the inline SVG icons first — they're huge and contain `#color` codes
    // that would otherwise be mistaken for "# 122"-style chapter labels.
    const list = (await fragment(`/series/${seriesId}/full-chapter-list`)).replace(/<svg[\s\S]*?<\/svg>/g, "");
    const raw: Array<{ id: string; number: number }> = [];
    const seen = new Set<string>();
    // Labels look like "Chapter 385", "Episode 5", or "# 122".
    for (const m of list.matchAll(
      /href="https:\/\/weebcentral\.com\/chapters\/([A-Z0-9]{10,})"[\s\S]{0,600}?>\s*(?:Chapter\s+|Episode\s+|#\s*)([0-9]+(?:\.[0-9]+)?)\s*</g,
    )) {
      const id = m[1];
      if (seen.has(id)) continue;
      seen.add(id);
      raw.push({ id, number: parseFloat(m[2]) });
    }
    raw.sort((a, b) => a.number - b.number);

    const info: MangaInfo = {
      id: seriesId,
      title,
      type: longStrip ? "Manhwa" : "Manga",
      forceStrip: longStrip,
      chapters: raw.map((ch, i) => ({
        id: ch.id,
        title: `Chapter ${ch.number}`,
        number: ch.number,
        index: i,
        pageCount: 0,
        scanId: "",
      })),
    };
    await writeCache(MANGA_CACHE, key, info);
    return info;
  },

  async pages(_seriesId, chapterId): Promise<ReadChapter> {
    const key = `wc:pages:${chapterId}`;
    const hit = await readCache<ReadChapter>(CHAPTERS_CACHE, key, TTL);
    if (hit) return hit;
    const html = await fragment(`/chapters/${chapterId}/images?is_prev=False&current_page=1&reading_style=long_strip`);
    const urls: string[] = [];
    for (const m of html.matchAll(/src="(https:\/\/[^"]+\.(?:webp|jpg|jpeg|png|avif))"/gi)) {
      urls.push(m[1]);
    }
    const out: ReadChapter = {
      id: chapterId,
      title: "",
      scanId: "",
      pages: urls.map((url, i) => ({ id: `${chapterId}-${i}`, url, number: i, width: 0, height: 0, aspectRatio: 1 })),
    };
    await writeCache(CHAPTERS_CACHE, key, out);
    return out;
  },

  async related(): Promise<DiscoveryItem[]> {
    return [];
  },
};
